sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala

```diff
@@ -67,6 +67,27 @@ case class LogicalRelation(
     s"Relation ${catalogTable.map(_.identifier.unquotedString).getOrElse("")}" +
       s"[${truncatedString(output, ",", maxFields)}] $relation"
   }
+
+  override lazy val metadataOutput: Seq[AttributeReference] = relation match {
+    case _: HadoopFsRelation =>
+      val resolve = conf.resolver
+      val outputNames = outputSet.map(_.name)
+      def isOutputColumn(col: AttributeReference): Boolean = {
+        outputNames.exists(name => resolve(col.name, name))
+      }
+      // filter out metadata columns that have names conflicting with output columns. if the file
+      // has a column "_metadata", then the data column should be returned not the metadata column
+      Seq(FileFormat.FILE_METADATA_COLUMNS).filterNot(isOutputColumn)
+    case _ => Nil
+  }
+
+  override def withMetadataColumns(): LogicalRelation = {
+    if (metadataOutput.nonEmpty) {
+      this.copy(output = output ++ metadataOutput)
```

Comment:
In this copy, the underlying hadoopFsRelation remains same and so `dataSchema` field in HadoopFsRelation doesn't have the metadata columns. This breaks the rule SchemaPruning. That rule uses the `dataSchema` and ends up creating an invalid plan.
This will need some handling by either fixing the dataSchema in HadoopFsRelation or by fixing the rule.